packages/core/echo/echo-schema/src/automerge/automerge-doc-loader.ts

```diff
@@ -0,0 +1,286 @@
+          actualUrl: event.handle.url,
+        });
+        objectsToRebind.push(updatedObject);
+      }
+    }
+    return {
+      updatedObjectIds: inlineChangedObjects,
+      objectsToRebind,
+      createdObjectIds,
+      linkedDocuments,
+    };
+  }
+
+  private async _initDocHandle(ctx: Context, url: string) {
+    const docHandle = this._automerge.repo.find(url as DocumentId);
+    // TODO(mykola): Remove check for global preference or timeout?
+    if (getGlobalAutomergePreference()) {
+      // Loop on timeout.
+      while (true) {
+        try {
+          await warnAfterTimeout(5_000, 'Automerge root doc load timeout (AutomergeDb)', async () => {
+            await cancelWithContext(ctx, docHandle.whenReady(['ready'])); // TODO(dmaretskyi): Temporary 5s timeout for debugging.
+          });
+          break;
+        } catch (err) {
+          if (`${err}`.includes('Timeout')) {
+            log.info('wraparound', { id: docHandle.documentId, state: docHandle.state });
+            continue;
+          }
+
+          throw err;
+        }
+      }
+    } else {
+      await asyncTimeout(docHandle.whenReady(['ready']), 1_000, 'short doc ready timeout with automerge disabled');
+    }
+
+    if (docHandle.state === 'unavailable') {
+      throw new Error('Automerge document is unavailable');
+    }
+
+    return docHandle;
+  }
+
+  private _createContextBoundSpaceRootDocument(ctx: Context) {
+    const docHandle = this._automerge.repo.create<SpaceDoc>();
+    this._spaceRootDocHandle = docHandle;
+    ctx.onDispose(() => {
+      docHandle.delete();
+      this._spaceRootDocHandle = null;
+    });
+  }
+
+  private _initDocAccess(handle: DocHandle<SpaceDoc>) {
+    handle.change((newDoc: SpaceDoc) => {
+      newDoc.access = { spaceKey: this._spaceKey.toHex() };
+    });
+  }
+
+  private async _createObjectOnDocumentLoad(handle: DocHandle<SpaceDoc>, objectId: string) {
+    try {
+      await handle.doc(['ready']);
+      const logMeta = { objectId, docUrl: handle.url };
+      const listener = this._documentLoadingListener;
+      if (listener == null) {
+        log.warn('document loaded after a listener was removed, ignoring', logMeta);
+        return;
+      }
+      const objectDocHandle = this._objectDocumentHandles.get(objectId);
+      if (objectDocHandle?.url !== handle.url) {
+        log.warn('object was rebound while a document was loading, discarding handle', logMeta);
+        return;
+      }
+      listener(handle, objectId);
+    } catch (err) {
+      const shouldRetryLoading = this._documentLoadingListener != null;
+      log.warn('failed to load a document', {
+        objectId,
+        automergeUrl: handle.url,
+        retryLoading: shouldRetryLoading,
+        err,
+      });
+      if (shouldRetryLoading) {
+        await this._createObjectOnDocumentLoad(handle, objectId);
+      }
+    }
+  }
+
+  private _getInlineAndLinkChanges(event: DocHandleChangePayload<SpaceDoc>) {
```

Comment:
if this method isn't using any class fields/methods it could be extracted as a helper function in this file